ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+through a clawback operation -- which results in a process that is opaque to
+users and wallets and a bad experience for the users.
+
+The Smart ASA defined by this ARC extends the ASA to increase its expressiveness
+and its flexibility. Developers can now introduce additional business logic
+around its operations. Wallets, dApps, and SDKs can recognize Smart ASAs and
+adjust their flows and user experiences accordingly.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC
+2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+The following sections describe:
+
+- The ABI interface for a controlling Smart Contract (the Smart Contract that
+  controls a Smart ASA).
+- The metadata required to denote a Smart ASA and define the association between
+  an ASA and its controlling Smart Contract.
+
+### ABI Interface
+
+The ABI interface specified here draws inspiration from the transaction
+[reference](https://developer.algorand.org/docs/get-details/asa/#asset-functions)
+of an Algorand Standard Asset (ASA).
+
+To provide a unified and familiar interface between the Algorand Standard Asset
+and the Smart ASA, method names and parameters have been adapted to the ABI
+types but left otherwise unchanged.
+
+#### Asset Creation
+
+```json
+{
+  "name": "asset_create",
+  "args": [
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "uint64" }
+}
+```
+
+Calling `asset_create` creates a new Smart ASA and returns the identifier of the
+ASA. The [metadata section](#metadata) describes its required properties.
+
+> Upon a call to `asset_create`, a reference implementation SHOULD:
+>
+> - Mint an Algorand Standard Asset (ASA) that MUST specify the properties
+>   defined in the [metadata section](#metadata). In addition:
+>   - The `manager`, `reserve` and `freeze` addresses SHOULD be set to the
+>     account of the controlling Smart Contract.
+>   - The remaining fields are left to the implementation, which MAY set `total`
+>     to `2 ** 64 - 1` to enable dynamically increasing the max circulating
+>     supply of the Smart ASA.
+>   - `name` and `unit_name` MAY be set to `SMART-ASA` and `S-ASA`, to
+>     denote that this ASA is Smart and has a controlling application.
+> - Persist the `total`, `decimals`, `default_frozen`, etc. fields for later
```

Comment:
There are a number of places in this spec that talk about persisting information:

- These parameters (`total`, `decimals`, etc.)
- The `manager_addr` for asset config
- `asset_freeze` (recommends global)
- `account_freeze` (recommends local)

If you are controlling multiple ASAs (e.g. hundreds / thousands) from the same contract (which I think is an important implementation option) then this is non-trivial to implement. I think there should be a section about how to handle multiple assets that talks through this. #bringonboxstorage

Reply:
True. 

I think that section should NOT be normative, but I can't write it before the design around boxes crystallises.

One thing we could write is how to handle multiple ASAs that share the same configuration -- meaning that require constant storage.

Reply:
Makes sense.

As a suggestion, can we at least have a section that discusses the fact that this standard is intended to support issuing multiple ASAs, but aspects of that implementation may be harder to implement in lieu of box storage given the limited amount of global storage available? That way it's clear to people this standard is open to supporting that pending people figuring out how to implement it in the future.